system/heap.c

```diff
@@ -0,0 +1,128 @@
+// SPDX-License-Identifier: GPL-2.0
+// Copyright (C) 2022 Martin Whitaker.
+
+#include <stdint.h>
+
+#include "boot.h"
+
+#include "memsize.h"
+#include "pmem.h"
+
+#include "heap.h"
+
+//------------------------------------------------------------------------------
+// Types
+//------------------------------------------------------------------------------
+
+typedef struct {
+    int         segment;
+    uintptr_t   start;
+    uintptr_t   end;
+} heap_t;
+
+//------------------------------------------------------------------------------
+// Private Variables
+//------------------------------------------------------------------------------
+
+static heap_t lm_heap = { .segment = -1, .start = 0, .end = 0 };
+static heap_t hm_heap = { .segment = -1, .start = 0, .end = 0 };
+
+//------------------------------------------------------------------------------
+// Private Functions
+//------------------------------------------------------------------------------
+
+static size_t num_pages(size_t size)
+{
+    return (size + PAGE_SIZE - 1) >> PAGE_SHIFT;
+}
+
+static uintptr_t heap_alloc(const heap_t *heap, size_t size, uintptr_t alignment)
+{
+    if (heap->segment < 0) {
+        return 0;
+    }
+    uintptr_t addr = pm_map[heap->segment].end - num_pages(size);
+    addr &= ~((alignment - 1) >> PAGE_SHIFT);
+    if (addr < heap->start) {
+        return 0;
+    }
+    pm_map[heap->segment].end = addr;
+    return addr << PAGE_SHIFT;
+}
+
+static uintptr_t heap_mark(const heap_t *heap)
```

Comment:
It seems fine, but I don't really understand the logic between `heap_mark` and `heap_rewind`.

Reply:
If no keyboards are found when probing the ports of a USB controller, that controller is reset and the memory allocated to it can be freed, The heap allocation scheme is very simple - it is implemented as a stack - so to free memory, you have to tell it how much memory to free. Rather than have the USB drivers track how much memory they have allocated, I thought it easier to record the level of the heap before allocating any memory, and then to rewind to that level if the allocated memory wasn't required.